pandas/tests/groupby/test_custom_metadata.py

```diff
@@ -0,0 +1,81 @@
+import numpy as np
+import pytest
+
+import pandas as pd
+
+
+class SubclassedDataFrame2(pd.DataFrame):
+    """
+    Extension of DataFrame as described in [guidelines]
+
+    [guidelines]: https://pandas.pydata.org/pandas-docs/stable/development/extending.html#override-constructor-properties  # noqa
+    """
+
+    # normal properties
+    _metadata = ["added_property"]
+
+    @property
+    def _constructor(self):
+        return SubclassedDataFrame2
+
+
+@pytest.mark.xfail(
+    reason="Missing high-performance implementation of metadata support for groupby. "
```

Comment:
what is the issue here?

Reply:
TLDR: I plan to submit a separate PR to lift this x-fail which is unrelated to the linked issue.

The initial commit in this PR (7cc4d5375e57277f316c5a8bb17a58ab8003c5f1) included a call to `__finalize__` in the concrete `DataSplitter`-classes, but as pointed out by @TomAugspurger  and @jbrockmendel , this led to performance-degradation for aggregation code-paths due to unneeded calls to `__finalize__` for the intermediate dataframe/series objects. This is what is tested for in the performance test added in a2bdc3d3dfe80018bb6cabce3583fdf9df6cdca9.

Since calling `__finalize__` in `DataSplitter` turned out to be irrelevant for the linked issue (which was fixed by adding a `__finalize__` call in `BaseGroupBy._agg_general`, see 8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e) it has been removed from this PR.

One solution to lift the x-fail would be to call `__finalize__` for the groups returned by the iterator returned by `BaseGroup.get_iterator` (see 0af015ba77d8ff05f0d38a3d4b0518bbc737d393 for an implementation on my fork). I am getting some spurious test failures on this branch.

The reason I have not included this solution in the present PR is that I would rather submit it in a separate PR: Calling `__finalize__` independently from object initialization could potentially lead to duplicate calls to `__finalize__` for the same object, and I do not have sufficient overview of the code so see if this is the case in the proposed implementation. My thought was that a separate PR would make it easier to get an overview of such issues -- but I would of course be happy to merge into this PR if you prefer.